crates/bevy_ui/src/interaction.rs

```diff
@@ -0,0 +1,222 @@
+use crate::Node;
+use bevy_app::{EventReader, Events};
+use bevy_core::FloatOrd;
+use bevy_ecs::prelude::*;
+use bevy_input::{mouse::MouseButton, Input};
+use bevy_math::{vec2, Vec2};
+use bevy_transform::components::Transform;
+use bevy_window::CursorMoved;
+use std::collections::HashSet;
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum PropagatePolicy {
+    Block,
+    Pass,
+}
+
+impl Default for PropagatePolicy {
+    fn default() -> Self {
+        PropagatePolicy::Block
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum Interaction {
+    Pressed,
+    Hovered,
+    None,
+}
+
+impl Default for Interaction {
+    fn default() -> Self {
+        Interaction::None
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum RegionAction {
+    Enter,
+    Exit,
+    Hover(Vec2),
+    Move(Vec2),
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum PressAction {
+    Up,
+    Down,
+}
+
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct PointerRegion {
+    pub entity: Entity,
+    pub action: RegionAction,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct PointerPress {
+    pub entity: Entity,
+    pub action: PressAction,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub struct PointerClick {
+    pub entity: Entity,
+}
+
+pub struct EventState {
+    cursor_moved_event_reader: EventReader<CursorMoved>,
+    cursor_position: Vec2,
+    hovered_entities: HashSet<Entity>,
+    pressed_entities: HashSet<Entity>,
+}
+
+impl Default for EventState {
+    fn default() -> Self {
+        EventState {
+            cursor_moved_event_reader: Default::default(),
+            cursor_position: Default::default(),
+            hovered_entities: HashSet::new(),
+            pressed_entities: HashSet::new(),
+        }
+    }
+}
+
+pub fn ui_event_system(
+    mut state: Local<EventState>,
+    mouse_button_input: Res<Input<MouseButton>>,
+    cursor_moved_events: Res<Events<CursorMoved>>,
+    mut region_events: ResMut<Events<PointerRegion>>,
+    mut click_events: ResMut<Events<PointerClick>>,
+    mut press_events: ResMut<Events<PointerPress>>,
+    mut node_query: Query<(
+        Entity,
+        &Node,
+        &mut Interaction,
+        &Transform,
+        Option<&PropagatePolicy>,
+    )>,
+) {
+    let mut cursor_has_moved = false;
+    if let Some(cursor_moved) = state.cursor_moved_event_reader.latest(&cursor_moved_events) {
+        state.cursor_position = cursor_moved.position;
+        cursor_has_moved = true;
+    }
+
+    let mut new_hovered_entities = HashSet::new();
+    {
+        let mut query_iter = node_query.iter();
+        let mut moused_over_z_sorted_nodes = query_iter
+            .iter()
+            .filter_map(|(entity, node, interaction, transform, propagate_policy)| {
```

Comment:
Avoid Multi-Line Closures. In this case, it makes the code harder to read, especially if we add another operation after the filter_map. It’s difficult to understand what the closure is doing upon first glance. Also, there’s a good chance that we will want to reuse the behavior of the closure elsewhere in our program. Simple extracting to a named function can help. And while doing that consider refactoring ui_event_system for better readability. Looking at the code we can see 10 if statements, one of which check more than one condition. It helps naming the conditions by extracting them to a private functions. Also great way of reducing the size of ui_event_system could be by extracting the if/else statement part into a smaller private functions.